### ISSUE TITLE
General questions about the dataset

### ISSUE THREAD

**Issue by Almo1982** [mcordts/cityscapesScripts#8]
Hi all,

i already trained some nets with your released dataset. everything worded fine so far for the training process. But i have a couple of questions if i understand everything correctly:

1. On your homepage it says there are 30 classes in 8 categories, so far so good. If i take a look into the labels.py file i count 35 classes (line 63 to 97). For me its not really clear what number to set as the number of output of my last conv layer. acutally after some experementing my suggestion is to set the number of outputs to 34. The problem is, i dont know exactly why... :-). 

2. I saw some net architectures where there is a "ignore label" parameter. Regarding your dataset my understanding is that the void-category should be ignored (better sayed, void is not included for evaluation). How to handle this? on your Homepage void category contains "ground", "dynamic" and "static", so should i set ignore_labesl to 3? in labels.py there are 7 labels within the void-category. This is a little bit confusing for me.

3. If i have 'class_weighting' by 'weight_by_label_freqs: true' in a "SoftmaxWithLoss"-Layer, how are the weights calculated for the cityscapes labels?

4. Could you explain the iIoU metric in a little bit more "easy" detailed way than described in your official paper? I still have some understanding issues here.

Thank you very much in advance for your help!

Regards
Stefan

**Comment by mcordts** [comment]
Hi Stefan,

thank you for your feedback. Regarding your items:

1. The 30 classes is more the high-level view of the dataset and reflects the number of manually labeled classes. The `labels.py` includes additional artificial classes such as ego-vehicle that can be labeled automatically with known camera calibration. For evaluation, we use only 19 of these classes, since we excluded the void category and rare cases. In `labels.py`, the excluded classes are flagged by `ignoreInEval=true`. Thus, I would set the number of output channels to 19, and treat pixels with one of the remaining classes in ground truth as ignore. 

2. I would use the IDs that we named `trainId` for training. The idea is to map labels that are ignored to 255 and to have consecutive IDs for all other labels. However, please make sure to reverse this mapping before evaluation.

3. In our experiments in the paper, we did not use any class weighting.

4. The iIoU weights pixels by the size of the ground truth object that they belong to. In doing so, objects of different pixel size (e.g. close and far cars) are treated more similar in evaluation. Maybe the implementation of the metric helps to understand this better, please refer to `evaluation/evalPixelLevelSemanticLabeling.py`

Best,
Marius

**Comment by Almo1982** [comment]
Hi Marius,

thank you very much for your Response. Please see my further questions below:

1. Ok, ist much more clear now for me, but when i set Output to 19 in my last conv layer and ignore_label to 255 in the softmax layeri get an "core dumped error" like "free(): invalid next size ...". i recognized it has to do with the number of Outputs. if i Change the number of Outputs to 34 it starts Training the Network. It works with 35 also, this is Kind of strange in my eyes. For now i have bad Evaluation results, seems to be because of the number of Outputs?! but i cant even start a Training because of the error when i set it to 19. Can you help here too?

2. ok, understand so far.

3. ok

4. ok, thx, i didnt even recognized the evalPixelLevelSemanticLabeling.py script :-P So far, I was just evaluating visually by comparing the predictions with the ground truth pics.

**Comment by mcordts** [comment]
Hi Stefan,

you should use ground truth images that match your channel numbers, i.e. all pixel values must be within 0...numberChannels-1 plus the ignore value (e.g. 255). We suggest to use the `trainIDs` for this, you can generate those with the tool `preparation/createTrainIdLabelImgs.py`.

Best,
Marius

**Comment by Almo1982** [comment]
Hi Marius,

ok, thank you, i will try this tomorrow, that might be the problem that i didnt train with the prepared data created by preparation/createTrainIdLabelImgs.py!

A new question came on tonight...i was trying to run evalPixelLevelSemanticLabeling.py for some actual trained data, just to know how the script works. It took my 2 hours and i was not able to run this correctly...i set the environment variable many times, checked the path by echo $CITYSCAPES_RESULTS, this path points to my result images with the naming like described in the script! i always get the error 
"ERROR: Found no prediction for ground truth /home/USERNAME/Downloads/CityscapesDataset/cityscapesscripts/evaluation/../../gtFine/val/frankfurt/frankfurt_000000_011810_gtFine_labelIds.png" I replaced my username here, just for information.

i really dont understand why it alwyas points to the gtFine/val/frankfurt folder??? My environment path is set correctly to another path...i tried to unset all environment variables again and created the result folder in ../../results...it still points to the frankfurt thing. my custom path isnt respected at all, no matter which version of the description in the script i try!

if i give a path or a file as input argument the script runs but it says "Evaluating 0 pairs of images..." and all the rest is "nan" because of dividing by zero as i understand. 

what am i doing wrong here??? i am out of ideas...

thank you again in advance for your great help!

Regards
Stefan

**Comment by Almo1982** [comment]
Hi Marius,

one more question came up regarding to question 1. from the first post:

i converted my images via createTrainIdLabelImgs.py, that worked so far. But when i set number of outputs in my last conv layer to 19 and ignore_label to 255, it says "CUBLAS_STATUS_SUCCESS (11 vs. 0)  CUBLAS_STATUS_MAPPING_ERROR".

I have to set the number of outputs to 255, then it starts training...is this behaviour correct? that would mean all the accuracy values on the caffe output from 20-254 dont care? 

thank you for your help!

Regards
Stefan

**Comment by mcordts** [comment]
Hi Stefan,

regarding your evaluation problem. The script prints the ground truth filename when it cannot find an appropriate result file. Is there a result file for each of the validation images somewhere within the folder "CITYSCAPES_RESULTS"? The files must be named frankfurt_000000_011810*.png or similar.

Did you double-check that your ground truth images contain only values between 0...18 or 255 and that your network is actually using these images? 

Best,
Marius

**Comment by Almo1982** [comment]
Hi, 

How Do i Check how many labels are in the gt picture? I thought the Scripts' Output produces the correct number of labels? 

Thanks, 
Stefan

**Comment by Almo1982** [comment]
Hi Marius,

ok, now i wrote a small script to read out the pixel values of the whole gt Image!

there are a few values of 128, 129 and 64 in the pictures...There seems to be the problem!

All the rest is within the range of 0-19 and the white areas are 255, what looks good so far.

But how can it happen that there are values of 128,129 and 64 inside the picture? I used your script to create the labelTrainId-Images...

Thanks
Stefan

**Comment by Almo1982** [comment]
To be a little bit more precise...i checked about 15 gt labelTrainId-Images now, the "strange" values always appear before or after a region of pixels with value 255 if this helps anyway?! I saw some additional "out of range"values 65, 192 and 134...but like i said, always before or after a 255-region.

thanks

**Comment by mcordts** [comment]
Hi Stefan, 
are you resizing the images with some kind of interpolation? That would now be a good idea for images with IDs.
Best,
Marius

**Comment by Almo1982** [comment]
Hi Marius,

thank you for your quick response!

In fact i do this by cv.Resize(image1,image2,interpolation=cv.CV_INTER_LINEAR) where image1 is the original size and image to is my new size (512x256). So there might be the problem? But i resized the images BEFORE i used your script createTrainIdLabelImgs.py! Should i create the trainId-Images with the original cityscapes image-sizes and then resize them or shouldn't i resize the images at all?

thank you!
Stefan

**Comment by Almo1982** [comment]
i checked the non-resized TrainId-images, no "strange" values in there, that seems to be the problem!!! 

Thank you very much for your support, i would have searched for this forever :-P
i think i can proceed now and get some results! Thank you!

Stefan

**Comment by Almo1982** [comment]
but is it possible to resize images now? I mean resizing the original labelTrainId images? i dont want to use this big resolution. Or will it corrupt the pixel values again?

**Comment by mcordts** [comment]
Just use INTER_NEAREST...
Good luck,
Marius

**Comment by Almo1982** [comment]
That worked! The net is training right now with image size 512x256, number of output 19 and ignore_label 255!

Thank you again very much!!!
Have a nice weekend!

Stefan

**Comment by mcordts** [comment]
Great. Thank you, and good luck with your training then. I will close the issue.
Best,
Marius

**Comment by JoltonDsouza** [comment]
Hi Maurius, @mcordts I want to use the cityscape dataset but want to reduce the classes to just 2(two) classes. I see that there are .json files I need to change but apart from that, to make a segmentation model, what more am I supposed to change?
Thanks

**Comment by mcordts** [comment]
If I understood you correctly, this is what you want to do:

1. In [labels.py](https://github.com/mcordts/cityscapesScripts/blob/master/cityscapesscripts/helpers/labels.py) you want to change the trainId column to the two classes that you are interested in (values 0 and 1). The remaining are typically mapped to 255 and then ignored during training (something you will need to take care of)
2. Run the script [csCreateTrainIdLabelImgs ](https://github.com/mcordts/cityscapesScripts/blob/master/cityscapesscripts/preparation/createTrainIdLabelImgs.py)
3. Train your model on the pngs produced by the previous step.

**Comment by JoltonDsouza** [comment]
> If I understood you correctly, this is what you want to do:
> 
> 1. In [labels.py](https://github.com/mcordts/cityscapesScripts/blob/master/cityscapesscripts/helpers/labels.py) you want to change the trainId column to the two classes that you are interested in (values 0 and 1). The remaining are typically mapped to 255 and then ignored during training (something you will need to take care of)
> 2. Run the script [csCreateTrainIdLabelImgs ](https://github.com/mcordts/cityscapesScripts/blob/master/cityscapesscripts/preparation/createTrainIdLabelImgs.py)
> 3. Train your model on the pngs produced by the previous step.

Exactly what I needed! Thanks 👍

**Comment by mcordts** [comment]
As you set the remaining pixels to 255, both 0 and 1 will appear black for the eye.

**Comment by herleeyandi** [comment]
Hello I just trying to use this dataset. Like @mcordts said it better to use `TrainIds` however after I download the dataset in file `gtFine_trainvaltest.zip (241MB)` I can't find it, instead I just find `labelIds`. Where can I download it?, or in which file it is located?
![image](https://user-images.githubusercontent.com/2309538/49430129-32674400-f7e5-11e8-9b86-69a5e1d5fb93.png)

**Comment by herleeyandi** [comment]
@mcordts I have tried to run the script `createTrainIdLabelImgs.py` but looks like some validation dataset's images are not converted well. This image is the example:
```
./gtFine/val/lindau/lindau_000035_000019_gtFine_labelIds.png
label : [ 1  3  4  6 11 17 20 21 24 26]
```
why it is not encoded to 0-18?

**Comment by RicheyHuang** [comment]
@mcordts Hi Maurius, are the classes of 'static' and 'unlabeled' necessary if I want to customize my own classes? Thank you in advance!

**Comment by Ashutosh1995** [comment]
If I want to train on the 8 categories, how could I do that ?
